mlir/lib/Dialect/PDL/IR/Builtins.cpp

```diff
@@ -39,6 +53,61 @@ mlir::Attribute addElemToArrayAttr(mlir::PatternRewriter &rewriter,
   values.push_back(element);
   return rewriter.getArrayAttr(values);
 }
+
+LogicalResult add(mlir::PatternRewriter &rewriter, mlir::PDLResultList &results,
+                  llvm::ArrayRef<mlir::PDLValue> args) {
```

Comment:
As long as I am fan of using this signature so that we can register it in both rewrite and matching section, I am wondering what you think about changing this to for better readability and register it using lambda:
```suggestion
LogicalResult add(mlir::PatternRewriter &rewriter, mlir::Attribute lhs, mlir::Attribute rhs) -> mlir::Attribute {
```

The only concern that I have is if this is going to be used in the rewrite, what should we return in case of failure() because if we return failure, it means the rewrite failed which is not supported by pdl_interp. However, at the same time, it's a rewrite failure to me.

Reply:
so this is also not important then based on the other comment of Matthias about upstreaming.